### ISSUE TITLE
Some linalg operations are not taking advantage from batched computation

### ISSUE THREAD

**Issue by o-alexandre-felipe** [pytorch/pytorch#64058]
## 🚀 Feature

The non linear matrix functions, in special the ones implemented in linalg should take advantage of the batch parallelism when possible.

## Motivation

Matrix are the representations for linear operations by excelence, matrix algebra is widely known so it makes easier to communicate, modelling and manipulations. When I talk use the seemingly paradoxical term nonlinear linear algebra functions I am referring to a few functions I tested from the linear algebra functions that performs differently than matrix multiplication, or elementwise operations. For instance inverse `inv(A+B) != inv(A) + inv(B)`, `qr(A+B) != qr(A) + qr(B)`, `eig(A+B) != eig(A) + eig(B)` among others.

I reported [a related issue](https://github.com/pytorch/pytorch/issues/63992) demonstrating that there is room for improvement for batched inversion of small matrices. I suggested there the use of an analytic formula for the 2 x 2 case. But I am not familiar with the codebase.

> For batches of 2x2 matrices [`apply_batched_inverse` from `aten/src/ATen/native/cuda/BatchLinearAlgebra.cu`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/BatchLinearAlgebra.cu#L1333) is used.
>
> The implementation in the backend library (Magma) is general that works for all sizes, there are no optimizations for small matrices.

_Originally posted by @IvanYashchuk in https://github.com/pytorch/pytorch/issues/63992#issuecomment-906244922_

Of course having a generic algorithm to handle all the cases is better than trying to handle every particular case (imagine a language without for loops), pointless. One possible interpretation of the above message is since "it uses a general method that works for all sizes" it cannot be improved. Another possible implementation is since "it uses a general method that works for all sizes" there are modifications that would improve the performance for all sizes.

The run time differences between the analytic formula and the generic implementation for 2 x 2 cannot be easily explained simply by the overhead of the generic algorithm, if I think about inversion with triangularization it consists in two row operations, and in the worse at most one row permutation. I decided to check with an implementation of a linear algorithm that works for all sizes how it performs.

## Pitch

The GPU demonstrated to successfully take advantage of batched operations in all sorts of computations, it makes me to believe that we can do the same with the matrix operations in the linalg module.

## Alternatives

I have created a [gist about batched nonlinear linalg functions](https://gist.github.com/o-alexandre-felipe/dedcbc88c89ed52601fb2dfb884a2608#file-linalg-qr-experiment-py-ipynb), there I implemented a generic batched QR decomposition in python using the matrix operations in pytorch and compared it with the native implementation.

## Additional context
The tests reveal that the linalg [QR decomposition](https://pytorch.org/docs/stable/generated/torch.linalg.qr.html#torch.linalg.qr), 
[eigenvalue decomposition](https://pytorch.org/docs/stable/generated/torch.linalg.eig.html#torch.linalg.eig), 
have the same throughput for batch sizes in the from 1 to 100k matrices of dimension 10 x 10. 
[Singular value decompositions](https://pytorch.org/docs/stable/generated/torch.linalg.svd.html#torch.linalg.svd) increase up to batch sizes 1k, 

![image](https://user-images.githubusercontent.com/19939799/131039541-5569b65f-fccd-4daf-96df-2876898be473.png)

[matrix inversion](https://pytorch.org/docs/stable/generated/torch.linalg.inv.html#torch.linalg.inv), becomes computation bounded at batch sizes of about 10k, the matrix multiplication and elementwise multiplications have basically the same performance.

![image](https://user-images.githubusercontent.com/19939799/131040876-3691663c-5f7b-4d34-98ed-40a3bc3f2459.png)

The matrix inversion requires about the number of operations required for two matrix multiplications, that's about the number of operations required by QR decomposition. For batches above 10k the python QR implementation overtake the linalg.inv. And the number of operations is probably not what is determining the run time. Check matrix multiplication and elementwise run-times, they follow very closely each other, and the of arithmetic operations for matrix multiplication was more than 10 times the number of operations required for elementwise multiplication.

cc @VitalyFedyunin @ngimel @heitorschueroff @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

**Comment by Lezcano** [comment]
I don't think it's too difficult or too easy, I think it just takes time. For each operation you need to implement it con C++ for CPU / CUDA and on CUDA you need to implement dispatches to cuSOLVER / MAGMA, with their batched and non-batched version. On CPU you are good, you just need to call LAPACK. After that it needs to be tested and debugged for all the possible flags that these functions take, and all the possible shapes and strides of tensors. Finally this needs to be benchmarked.

It can be done, and it will be done, but it takes time. We want to converge to a `torch.linalg` module that both has performances as one would expect, while having a simple and easy to use API and a good documentation. Now, we do accept contributions, so if you want to have a go at it, please feel more than welcome!

cc @IvanYashchuk as he was working on improving the performance of the QR decomposition a few months ago.
cc @xwang233

**Comment by o-alexandre-felipe** [comment]
@Lezcano could you check [this](https://gist.github.com/o-alexandre-felipe/dedcbc88c89ed52601fb2dfb884a2608#file-batch-1-computation-ipynb). This time I tested with a large matrix. One more indication that there is something wrong.

My expectation was that most of the linear algebra functions would get a speedup when running on the GPUs. Or the initiative of writing linalg for GPU was just for completeness and ease of use?

You guys that know the code from inside (specially the NVIDIA gurus), don't you think it is possible that some basic function that is called everywhere may be doing something that prevents parallelism, or may allocating memory in the innermost loop?

I ran that notebook in a google colab environment and it shows the same

![image](https://user-images.githubusercontent.com/19939799/131260022-47a3b7ba-dfe1-442e-8403-44a016b74344.png)

**Comment by nikhilvyas** [comment]
Has there been any update on this? QR decomposition has recently been used in second order optimizers (https://arxiv.org/pdf/2405.18144, https://arxiv.org/abs/2409.11321). For networks such as Transformers there are many layers on equal size, a batched version of QR which exploits gpu parallization would significantly reduce the overhead of these optimizers. This also applies to symmetric eigenvector decomposition (https://github.com/facebookresearch/optimizers/tree/main/distributed_shampoo) but I mentioned QR first since that is probably easier to parallelize. 

I am happy to expand on this use case if that would be helpful.

**Comment by lezcano** [comment]
This is already implemented in `torch.linalg`. Closing

**Comment by nikhilvyas** [comment]
Could you point me to the PR which implemented this? In my experiments there is no benefit to batching.

**Comment by lezcano** [comment]
can you post here a benchmark using PyTorch benchmarking tools that show that?

**Comment by nikhilvyas** [comment]
Here is a simple test:

```
import torch

single_time_avg = 0
batch_time_avg = 0

num_iter = 40
start_iter = 10

for i in range(num_iter):
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    G = torch.randn(1024, 1024).cuda()
    start.record()
    _, _ =     Q, R = torch.linalg.qr(G)
    end.record()
    torch.cuda.synchronize()
    if i > start_iter:
        time_passed = (start.elapsed_time(end))
        single_time_avg += time_passed

    batch_size = 10
    G_batch =  torch.randn(batch_size, 1024, 1024).cuda()
    start.record()
    _, _ = torch.linalg.qr(G_batch)
    end.record()
    torch.cuda.synchronize()
    if i > start_iter:
        time_passed = (start.elapsed_time(end))
        batch_time_avg += time_passed


print('\nsingle time (avg)', single_time_avg/(num_iter-start_iter-1), flush=True)
print('batch time (avg)', batch_time_avg/(num_iter-start_iter-1), flush=True)
```

This gives

```
single time (avg) 4.743197549747515
batch time (avg) 46.75045583217959
```

which shows that have a batch size of 10 gives no benefits (i.e. it takes 10x longer).

**Comment by lezcano** [comment]
In that case, the batched function is just calling the non-batched function in a loop:
https://github.com/pytorch/pytorch/blob/8c860aef0d640fa18cc4e722af3a7f7b81eb2a16/aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp#L1849-L1855
and same for the part where you multiply the Householder reflections
https://github.com/pytorch/pytorch/blob/f42d1b6fa12743a8b5cb7abe96d5b368db014c5f/aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp#L1771-L1782
so this is somewhat expected. As you see in that note, the Householder part may be improved using magma, but no one is working on this ATM. On the cusolver side, this is as good as it gets, as cusolver does not expose a batched version of orgqr.

More generally, you'd be better off using `torch.utils.benchmark` to benchmark, as here you are synchronising after every call to the function, which is less than good. You are also not warming up, and you are not averaging across non-synchronized calls to reduce variance. In this case, those things are not very relevant, as most linalg functions synchronize for error checking, but generally, all these are already taken into account by `torch.utils.benchmarks`.

**Comment by nikhilvyas** [comment]
Got it, I will try out `torch.utils.benchmarks`.
